### ISSUE TITLE
[feature] Support for custom `ReadAt` implementations (e.g. for request cancellation)

### ISSUE THREAD

**Issue by josxha** [bramp/pmtiles-dart#3]
Hi Andrew, sorry for opening another issue. I tried to add support for request cancellation for `flutter_map_pmtiles` but couldn't because of some limitations of this package.

### Use case
If we want to display tiles on a map (e.g. on flutter_map) we request a decent amout of tiles. If the viewport of the map changes quickly, tiles are potentially not required any more if they finish loading. Not being able to cancel those obsolete requests leads to a queue of requested tiles on web and delays the loading of new tiles.

(Note: I throttled the network speed in this video for make this effect more visible.)

https://github.com/bramp/pmtiles-dart/assets/34318751/f589f788-bfa2-4ef4-ac64-3f5b19e93317

I assume that request cancellation is out of scope for this package since `http` has no support for it. However, it would be still interessting to open the API of this package to allow such implementations externally. 

### Potential solution

The http integration is currently located in the `HttpAt` class that implements `ReadAt`. If the package allows users to provide their own `ReadAt` implementations, they could use for example `dio` to handle the requests that has support for request cancellation.

```dart
final Response<ResponseBody> response = await dio.get(url,
    cancelToken: myCancelToken,
    options: Options(
      responseType: ResponseType.stream,
      headers: {
        HttpHeaders.rangeHeader: 'bytes=$offset-${offset + length - 1}',
      },
    ));
```

There are however currently some limitations in the `pmtiles` package that prevents this.

### Current limitations

1) The `PmTilesArchive.fromReadAt()` constructor is currently marked to be only visible for testing. This "prevents" (shows a warning) to create an PmTilesArchive from a custom ReadAt implementation.
2) There is no internal way to pass CancelTokens through to ReadAt. This can be bypassed by storing all tokens externally in a `Map<int, CancelToken>` where the keys are the tileIds. But the missing link is that the tileId isn't provided to `ReadAt.readAt()` and therefore ReadAt has no information which CancelToken to use. This could be solved if `List<int>` of requested tileIds gets added as a parameter to `.readAt()`.

**Comment by bramp** [comment]
Thanks for the feature request. Just quickly reading though, 

1. Should be no issue to make public

2. I'm not sure we want to pass tile IDs down into the IO layer.  That leaks abstractions, and not all ReadAt are for tiles, sometimes we just want metadata, etc.

This is also made more complex, that a read may be for many tiles, and you may just want to cancel a handful of them. Do we cancel the whole request? Do you want to cancel and then re-issue for a subset?

I'd have to look into this more to come up with a plan.  The `PmTilesArchive.tiles` method does return a `Stream<Tile>`, and Streams can be cancelled/closed. So that at least provides a signal down to the IO layer that it should be ended. We could also swap out the `Futures`, with `CancellableFuture` (from https://pub.dev/packages/cancellation_token)

**Comment by josxha** [comment]
> I'm not sure we want to pass tile IDs down into the IO layer. That leaks abstractions, and not all ReadAt are for tiles, sometimes we just want metadata, etc.

Agree, the implemented abstraction of the package is really nice and it would be a shame to break these abstraction layers.

> Do we cancel the whole request? Do you want to cancel and then re-issue for a subset?

That's indeed a valid concern. As far as I'm aware of the tiles get returned directly as a Stream?
For my understanding it makes sense to let the ReadAt implementation decide on how (and if) to handle the cancellation of the request? It would need to know about what range of the initially requested range has been cancelled.

It could then decide whether
- it doesn't react to the cancellation
- the request gets cancelled and a cancellation exception gets thrown
- the request gets cancelled and some default data gets returned
- all tiles that are not cancelled are fetched and it can cancel the rest of the request and return

An questionable case would be how it would handle a cancellation of tiles that are in the middle of the initial requested range.

> and Streams can be cancelled/closed.

That's true but I had the information that it is (or maybe was) not possible to detect if the Stream has been cancelled and trigger some code? Maybe my knowledge is outdated here and this is possible now.

> We could also swap out the Futures, with CancellableFuture

The function that the package provides sounds good, and it would probably be a elegant implementation. Although it would need another dependency to be added to the package.

Some alternative idea would be a similar implementation to what flutter_map makes use of: A Future gets passed down to the tile provider and the Future returns when the tile request should get cancelled (a callback would probably work too and could return information about what action it took). The advantage of this implementation is that it doesn't rely on other dependencies. 

In the ReadAt (using dio) it would be possible to use it simply with:
```dart
// cancel the dio CancelToken when the cancelLoading Future returns
cancelLoading.then((_) => cancelToken.cancel());
```

---
I'm excited what solution you'll end up using - and thanks a lot looking into this!
